### ISSUE TITLE
(z3str3) Invalid model for formula with Strings and BitVecs

### ISSUE THREAD

**Issue by wolffdy** [Z3Prover/z3#4469]
Hi, I found the following formula gives an invalid model when using the z3str3 solver:

[bug.smt2.gz](https://github.com/Z3Prover/z3/files/4691709/bug.smt2.gz)
```
%  z3 model_validate=true smt.string_solver=z3str3 bug.smt2 
sat
(error "line 8 column 10: an invalid model was generated")

% cat bug.smt2
(declare-const a (_ BitVec 8))
(declare-const c Int)
(declare-fun e () String)
(assert (= (str.at e c) (seq.unit a)))
(assert (= 16 (str.len e)))
(assert (= (str.at e 5) (seq.unit #x00)))
(assert (= (= (str.indexof (str.substr e 0 (- 16 0)) (seq.unit #x00)) (- 1)) false))
(check-sat)
```
Verified on master, commit: 94ffd63
OS: Ubuntu 18.04

**Comment by mtrberzi** [comment]
In testing this formula I observed that this is unfortunately not a well-sorted formula. `(seq.unit #x00)` is not shorthand for a character constant, but in fact it has the sort `(Seq (_ BitVec 8))`. This isn't compatible with `String` (and especially now that Z3str3 supports Unicode, it cannot be assumed that it could always be made compatible!). The latest version of Z3 actually reports this as an error right now because the sorts are incompatible.

**Comment by NikolajBjorner** [comment]
The types were updated. 

```
(declare-const a Unicode)
(declare-const c Int)
(declare-fun e () String)
(assert (= (str.at e c) (seq.unit a)))
(assert (= 16 (str.len e)))
(assert (= (str.at e 5) (seq.unit (_ Char 0))))
(assert (= (= (str.indexof (str.substr e 0 (- 16 0)) (seq.unit (_ Char 0))) (- 1)) false))
(check-sat)
```
C:\z3\build>z3 4469.smt2 smt.string_solver=z3str3
ASSERTION VIOLATION
File: ..\src\smt\theory_str.cpp
Line: 212
!m.is_true(_e)

**Comment by mtrberzi** [comment]
Maybe the right way to handle this is to ensure that the argument to `seq.unit` has the same sort as whatever the current character sort is, and then handle it in the same way as character constraints are handled currently.

**Comment by mtrberzi** [comment]
Fixed in 04ac5f0
